File: ios/ParkingKok/Detection/Persistence/DiagnosticsReport.swift
import Foundation

/// The P0 diagnostics, flattened for export beside the checkpoint.
///
/// Why it exists: the motion samples, authorization statuses and drop counters that
/// prove the detection path worked lived only in memory and in `.info` log lines, so
/// reading them back from a device needed `sudo log collect`. Written as a file instead,
/// it comes off the device over the same path the checkpoint already uses, with no root.
///
/// **A hand-written projection, never an encoding of the live types.** `DetectionCheckpoint`
/// is `Codable` and carries `lastReliableLocation`, which holds real coordinates now that
/// M0A-2 selects one. This file is copied off the device by design, so encoding the
/// checkpoint wholesale would walk parking coordinates straight past
/// `docs/00_CORE_RULES.md` Privacy. Every field below is listed by hand; the
/// coordinate-bearing ones are deliberately absent, and a test on the encoded bytes
/// enforces it. **Adding a coordinate here is never the fix for a failing test.**
struct DiagnosticsReport: Sendable, Equatable, Codable {
    /// Bumped to 5 by the §7 distance-clause instrumentation; 4 was the movement-evidence
    /// counters before it.
    static let schemaVersion = 7

    var schemaVersion: Int = DiagnosticsReport.schemaVersion
    var generatedAt: Date

    // Launch and rehydration
    var launchReason: String
    var rehydratedAt: Date?
    var checkpointLoad: String
    var checkpointAge: TimeInterval?
    var isBeyondMotionRetention: Bool

    // Checkpoint, projected
    var state: String?
    var revision: Int?
    var stateEnteredAt: Date?
    var lastAutomotiveAt: Date?
    var lastLocationAt: Date?
    var travelDistanceEstimate: Double?
    var hasCandidate: Bool
    /// Presence and quality of the reliable fix — never where it was.
    var hasReliableLocation: Bool
    var reliableLocationCapturedAt: Date?
    var reliableLocationAccuracy: Double?

    // Motion
    var motionWindowStart: Date?
    var motionWindowEnd: Date?
    var motionSampleCount: Int
    var motionFailure: String?
    var motionSamples: [MotionSample]

    // Bounded driving session (docs/04 §3 DRIVING, docs/05 §7)
    var isCapturingDrivingLocation: Bool
    var drivingSessionStartedAt: Date?
    var drivingSessionCount: Int
    var drivingSessionResumedFromCheckpoint: Bool
    var drivingConfirmedAt: Date?
    var lastDrivingSessionEndReason: String?
    var lastDrivingSessionEndedAt: Date?
    var drivingFixCount: Int
    var drivingMovingSampleCount: Int
    /// docs/05 §7 movement evidence, instrumented. `speedMissingCount == fixCount` with
    /// `movingSampleCount == 0` is the exact signature of the defect these fields were
    /// added for: the speed-only rule could not confirm a drive underground.
    var speedAvailableCount: Int
    var speedMissingCount: Int
    var derivedMovingSampleCount: Int
    /// Why the distance fallback last declined a fix, or absent if it never has.
    var movementEvidenceRejectReason: String?
    var drivingOutlierDropCount: Int
    var drivingDistanceMeters: Double
    /// Legs §7's noise floor kept out of `drivingDistanceMeters`. Same name and meaning on
    /// Android, so one number can be compared across a pair of field runs.
    var distanceNoiseFloorRejectCount: Int
    /// How often a fix was accepted as `lastReliableLocation` — never which fix.
    var reliableLocationUpdateCount: Int
    var reliableLocationRejectCount: Int
    var lastReliableLocationRejection: String?
    var lastVehicleEvidenceAt: Date?
    var lastVehicleEvidenceConfidence: String?
    var captureFailure: String?

    // Trace recording (docs/05 §9). Counts only — the traces themselves are separate
    // files, and the point of these numbers is to tell whether recording is working
    // without retrieving any of them.
    var traceSessionCount: Int
    var traceEventCount: Int
    /// Sessions the rolling cap discarded. §9 requires the eviction to be visible; a
    /// silently shrinking history would look like recording that never happened.
    var traceDroppedSessionCount: Int
    /// Sessions discarded at rotation for holding one event or none (§9 "비생존 세션은
    /// 버린다"). Separate from the rolling-cap count on purpose: a rolling-cap eviction
    /// means the device is recording more than it can hold, while this climbing means the
    /// boundary is cutting sessions where there was nothing to cut — five of the nine
    /// September 2026 field sessions were single events. §9: "조용히 버리지 마라."
    var traceNonViableDropCount: Int
    var traceUnlabeledSessionCount: Int
    /// Closed sessions that were worth a label prompt and did not get one, because
    /// notifications are not permitted. The number that tells a field weekend with no
    /// labels apart from a field weekend with no travel. Same name on Android.
    var traceLabelPromptSuppressedCount: Int
    /// Sessions carrying a gap measurement, so the three numbers below can be read as a
    /// sample size rather than as a claim about every trace on disk.
    var traceMeasuredSessionCount: Int
    /// The §9 gap aggregate, over the sessions still on disk. **Instrumentation for a
    /// decision not yet taken**: the 30-minute idle gap came from a single day and missed
    /// by 66 seconds, and it does not move until these have accumulated.
    var traceMaxGapMillis: Int64
    var traceSessionsOver10MinGapCount: Int
    var traceSessionsOver20MinGapCount: Int
    /// Location observations refused as replays of an instant already recorded. Zero is
    /// the expected reading; a climbing count with a growing trace is Core Location
    /// re-delivering cached fixes and being ignored, which is the point.
    var traceReplayDropCount: Int
    var traceFailure: String?

    // Location quality
    var significantChangeCount: Int
    var lastLocationAccuracy: Double?
    var staleLocationDropCount: Int
    var lastStaleLocationAge: TimeInterval?
    var supersededLocationDropCount: Int
    var locationFailure: String?

    // Permissions and wiring
    var locationAuthorization: String
    var motionAuthorization: String
    var isMotionHistoryAvailable: Bool
    var isMonitoringSignificantChanges: Bool
    /// How long monitoring has been up, and how long since any detection input arrived.
    ///
    /// "Monitoring" and "receiving" are different claims and this file only made the
    /// first. A long age beside a longer silence is the shape worth suspecting; a short
    /// silence just means the pipeline is quiet.
    var monitoringAge: TimeInterval?
    var timeSinceLastDetectionInput: TimeInterval?
    var isSmartDetectionEnabled: Bool
    var storeSetupFailure: String?
    var lastPersistError: String?

    init(
        snapshot: RehydrationSnapshot,
        now: Date,
        locationAuthorization: LocationAuthorization,
        motionAuthorization: MotionAuthorization,
        isMotionHistoryAvailable: Bool,
        isMonitoringSignificantChanges: Bool,
        monitoringStartedAt: Date?,
        isSmartDetectionEnabled: Bool,
        storeSetupFailure: String?,
        traceSummary: TraceSummary
    ) {
        generatedAt = now

        launchReason = snapshot.launchReason.rawValue
        rehydratedAt = snapshot.rehydratedAt
        checkpointLoad = Self.describe(snapshot.checkpointLoad)
        checkpointAge = snapshot.checkpointAge
        isBeyondMotionRetention = snapshot.isBeyondMotionRetention

        let checkpoint = snapshot.currentCheckpoint
        state = checkpoint?.state.rawValue
        revision = checkpoint?.revision
        stateEnteredAt = checkpoint?.stateEnteredAt
        lastAutomotiveAt = checkpoint?.lastAutomotiveAt
        lastLocationAt = checkpoint?.lastLocationAt
        travelDistanceEstimate = checkpoint?.travelDistanceEstimate
        hasCandidate = checkpoint?.candidateId != nil
        hasReliableLocation = checkpoint?.lastReliableLocation != nil
        reliableLocationCapturedAt = checkpoint?.lastReliableLocation?.capturedAt
        reliableLocationAccuracy = checkpoint?.lastReliableLocation?.horizontalAccuracy

        motionWindowStart = snapshot.motionWindow?.start
        motionWindowEnd = snapshot.motionWindow?.end
        motionSampleCount = snapshot.motionSamples.count
        motionFailure = snapshot.motionFailure
        motionSamples = snapshot.motionSamples

        isCapturingDrivingLocation = snapshot.isCapturingDrivingLocation
        drivingSessionStartedAt = snapshot.drivingSessionStartedAt
        drivingSessionCount = snapshot.drivingSessionCount
        drivingSessionResumedFromCheckpoint = snapshot.drivingSessionResumedFromCheckpoint
        drivingConfirmedAt = snapshot.drivingConfirmedAt
        lastDrivingSessionEndReason = snapshot.lastDrivingSessionEndReason?.rawValue
        lastDrivingSessionEndedAt = snapshot.lastDrivingSessionEndedAt
        drivingFixCount = snapshot.drivingFixCount
        drivingMovingSampleCount = snapshot.drivingMovingSampleCount
        speedAvailableCount = snapshot.drivingSpeedAvailableCount
        speedMissingCount = snapshot.drivingSpeedMissingCount
        derivedMovingSampleCount = snapshot.drivingDerivedMovingSampleCount
        movementEvidenceRejectReason = snapshot.movementEvidenceRejectReason?.rawValue
        drivingOutlierDropCount = snapshot.drivingOutlierDropCount
        drivingDistanceMeters = snapshot.drivingDistanceMeters
        distanceNoiseFloorRejectCount = snapshot.drivingDistanceNoiseFloorRejectCount
        reliableLocationUpdateCount = snapshot.reliableLocationUpdateCount
        reliableLocationRejectCount = snapshot.reliableLocationRejectCount
        lastReliableLocationRejection = snapshot.lastReliableLocationRejection?.rawValue
        lastVehicleEvidenceAt = snapshot.lastVehicleEvidenceAt
        lastVehicleEvidenceConfidence = snapshot.lastVehicleEvidenceConfidence?.rawValue
        captureFailure = snapshot.captureFailure

        traceSessionCount = traceSummary.sessionCount
        traceEventCount = traceSummary.eventCount
        traceDroppedSessionCount = traceSummary.droppedSessionCount
        traceNonViableDropCount = traceSummary.nonViableDropCount
        traceUnlabeledSessionCount = traceSummary.unlabeledSessionCount
        traceLabelPromptSuppressedCount = traceSummary.labelPromptSuppressedCount
        traceMeasuredSessionCount = traceSummary.measuredSessionCount
        traceMaxGapMillis = traceSummary.maxGapMillis
        traceSessionsOver10MinGapCount = traceSummary.sessionsOver10MinGapCount
        traceSessionsOver20MinGapCount = traceSummary.sessionsOver20MinGapCount
        traceReplayDropCount = snapshot.traceReplayDropCount
        traceFailure = snapshot.traceFailure

        significantChangeCount = snapshot.significantChangeCount
        lastLocationAccuracy = snapshot.lastLocationAccuracy
        staleLocationDropCount = snapshot.staleLocationDropCount
        lastStaleLocationAge = snapshot.lastStaleLocationAge
        supersededLocationDropCount = snapshot.supersededLocationDropCount
        locationFailure = snapshot.locationFailure

        self.locationAuthorization = locationAuthorization.rawValue
        self.motionAuthorization = motionAuthorization.rawValue
        self.isMotionHistoryAvailable = isMotionHistoryAvailable
        self.isMonitoringSignificantChanges = isMonitoringSignificantChanges
        monitoringAge = monitoringStartedAt.map { now.timeIntervalSince($0) }
        // Any input counts: a motion edge, a significant change, or a bounded fix. The
        // question is whether the stack is hearing anything at all.
        timeSinceLastDetectionInput = [
            snapshot.motionSamples.last?.timestamp,
            snapshot.lastLocationAt,
            snapshot.lastVehicleEvidenceAt
        ]
        .compactMap(\.self)
        .max()
        .map { now.timeIntervalSince($0) }
        self.isSmartDetectionEnabled = isSmartDetectionEnabled
        self.storeSetupFailure = storeSetupFailure
        lastPersistError = snapshot.lastPersistError
    }

    private static func describe(_ load: DetectionCheckpointLoadResult) -> String {
        switch load {
        case let .restored(checkpoint): "restored(rev \(checkpoint.revision))"
        case .absent: "absent"
        case let .failed(failure): "failed(\(failure.diagnosticDescription))"
        }
    }
}

/// Writes the report beside the checkpoint so it inherits the same directory protection
/// class and comes off the device over the same path.
///
/// Best-effort by design: a diagnostics file that fails to write must never take down a
/// detection callback. The failure surfaces in the next report instead.
protocol DiagnosticsReportStoring: Sendable {
    func write(_ report: DiagnosticsReport) throws
}

struct FileDiagnosticsReportStore: DiagnosticsReportStoring {
    private static let fileName = "diagnostics.json"

    private let fileURL: URL

    init(fileURL: URL) {
        self.fileURL = fileURL
    }

    /// Beside `checkpoint.json`, so the directory's protection class covers both.
    static func defaultFileURL() throws -> URL {
        try FileDetectionCheckpointStore.defaultFileURL()
            .deletingLastPathComponent()
            .appending(path: fileName, directoryHint: .notDirectory)
    }

    func write(_ report: DiagnosticsReport) throws {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        encoder.dateEncodingStrategy = .iso8601
        let data = try encoder.encode(report)
        try data.write(to: fileURL, options: [.atomic])
    }
}

File: android/app/src/main/java/com/parkingkok/app/diagnostics/DiagnosticsReport.kt
package com.parkingkok.app.diagnostics

import com.parkingkok.app.detection.RegistrationStatus
import com.parkingkok.app.domain.detection.DetectionCheckpoint
import com.parkingkok.app.domain.detection.MotionDomainEvent
import com.parkingkok.app.domain.location.LocationDiagnosticsCounters
import com.parkingkok.app.domain.location.LocationQualityEntry
import com.parkingkok.app.domain.location.LocationSessionState
import com.parkingkok.app.domain.trace.TraceSummary
import kotlinx.serialization.Serializable

/**
 * The P0 diagnostics, flattened for export.
 *
 * Why it exists: the counters and session state that prove the detection path worked live
 * in DataStore, and reading them back meant scraping the preferences file with
 * `adb shell run-as ... | strings` — which shows whatever happens to be printable and
 * silently omits the rest. Written as a JSON file instead, the same evidence comes off the
 * device intact, still with no root. Same purpose as the iOS `DiagnosticsReport`.
 *
 * **A hand-written projection, never an encoding of the live types.** [DetectionCheckpoint]
 * is `@Serializable` and carries `lastReliableLocation`, which holds real coordinates from
 * M0B-2 onward. This file is copied off the device by design, so serializing the checkpoint
 * wholesale would walk parking coordinates straight past docs/00_CORE_RULES.md Privacy.
 * Every field below is listed by hand, the coordinate-bearing ones are deliberately absent,
 * and `DiagnosticsReportTest` enforces it against the encoded bytes rather than against the
 * field list — a redacted `toString` would not have caught it, because the encoder does not
 * call `toString`.
 */
@Serializable
data class DiagnosticsReport(
    val schemaVersion: Int = SCHEMA_VERSION,
    val generatedAtMillis: Long,

    // Permissions and wiring
    val activityRecognitionGranted: Boolean,
    val foregroundLocationGranted: Boolean,
    val backgroundLocationGranted: Boolean,
    val smartDetectionEnabled: Boolean,
    val transitionRegistration: String,

    // Checkpoint, projected
    val state: String?,
    val revision: Long?,
    val stateEnteredAtMillis: Long?,
    val lastAutomotiveAtMillis: Long?,
    val lastLocationAtMillis: Long?,
    val travelDistanceEstimateMeters: Double?,
    val hasCandidate: Boolean,
    /** Presence and quality of the reliable fix — never where it was. */
    val hasReliableLocation: Boolean,
    val reliableLocationCapturedAtMillis: Long?,
    val reliableLocationAccuracyM: Float?,

    // Bounded location session
    val sessionMode: String,
    val sessionStartedAtMillis: Long?,
    val sessionHardDeadlineAtMillis: Long?,
    val sessionRegistrationExpiresAtMillis: Long?,
    val sessionDeliveredUpdateCount: Int?,
    val sessionProfileVersion: Int?,
    val lastSessionStopReason: String?,
    val lastSessionFailure: String?,

    // §7 driving guard
    val drivingConfirmed: Boolean,
    val drivingReasonCodes: List<String>,
    val vehicleFirstSeenAtMillis: Long?,
    val lastVehicleEvidenceAtMillis: Long?,
    val reliableSampleCount: Int?,
    val maxSpeedMps: Float?,

    /**
     * §7's movement clause, as counts rather than as a verdict.
     *
     * The pair that has to be readable off one file: "confirmation never fired" and "no fix
     * ever carried a speed" look identical from outside, and on the September 2026 field
     * device it was the second one. [speedMissingCount] high with
     * [derivedMovingSampleCount] at zero is the shape of that defect; the same two with
     * [movementEvidenceRejectReason] set says which gate declined and why.
     */
    val movingSampleCount: Int?,
    val speedAvailableCount: Int?,
    val speedMissingCount: Int?,
    val derivedMovingSampleCount: Int?,
    val movementEvidenceRejectReason: String?,
    /** Fixes discarded as implausible jumps before the movement clause saw them (§5). */
    val movementOutlierCount: Int?,

    // Location quality counters (docs/05 §5: exclusions are counted, not dropped quietly)
    val counters: LocationDiagnosticsCounters,
    /** Volatile ring buffer, oldest first. Coordinate-free by construction. */
    val recentFixes: List<FixQuality>,

    // Motion transitions
    val transitionEventCount: Int,
    val recentTransitions: List<TransitionSummary>,

    /**
     * What the trace recorder has on disk
     * (docs/05_CROSS_PLATFORM_DOMAIN_CONTRACT.md §9).
     *
     * Here rather than only in the traces directory because this is the file a field run
     * is read from: "40 sessions, 12 discarded, 31 still unlabelled" is the difference
     * between a recording that is ready to convert and one that needs an evening of
     * labelling, and neither number is visible from a directory listing.
     */
    val trace: TraceSummary = TraceSummary(),
) {
    /** One ring-buffer entry, projected. */
    @Serializable
    data class FixQuality(
        val atMillis: Long,
        val accuracyM: Float,
        val speedMps: Float?,
        /** null when the fix was admitted. */
        val dropReason: String?,
    )

    /** One received transition, projected. */
    @Serializable
    data class TransitionSummary(
        val kind: String,
        val atMillis: Long,
        /** OEM delivery delay, the number docs/04_ANDROID_IMPLEMENTATION.md §20 asks for. */
        val deliveryDelayMillis: Long,
    )

    companion object {
        /** Bump whenever the shape changes, so an older payload is rejected, not half-read. */
        const val SCHEMA_VERSION: Int = 3

        @Suppress("LongParameterList")
        fun from(
            nowMillis: Long,
            checkpoint: DetectionCheckpoint?,
            sessionState: LocationSessionState,
            qualityHistory: List<LocationQualityEntry>,
            transitions: List<MotionDomainEvent>,
            activityRecognitionGranted: Boolean,
            foregroundLocationGranted: Boolean,
            backgroundLocationGranted: Boolean,
            smartDetectionEnabled: Boolean,
            transitionRegistration: RegistrationStatus,
            trace: TraceSummary,
        ): DiagnosticsReport {
            val record = sessionState.record
            val evidence = sessionState.evidence
            return DiagnosticsReport(
                generatedAtMillis = nowMillis,
                activityRecognitionGranted = activityRecognitionGranted,
                foregroundLocationGranted = foregroundLocationGranted,
                backgroundLocationGranted = backgroundLocationGranted,
                smartDetectionEnabled = smartDetectionEnabled,
                transitionRegistration = transitionRegistration.wire(),
                state = checkpoint?.state?.name,
                revision = checkpoint?.revision,
                stateEnteredAtMillis = checkpoint?.stateEnteredAtMillis,
                lastAutomotiveAtMillis = checkpoint?.lastAutomotiveAtMillis,
                lastLocationAtMillis = checkpoint?.lastLocationAtMillis,
                travelDistanceEstimateMeters = checkpoint?.travelDistanceEstimateMeters,
                hasCandidate = checkpoint?.candidateId != null,
                hasReliableLocation = checkpoint?.lastReliableLocation != null,
                reliableLocationCapturedAtMillis = checkpoint?.lastReliableLocation?.capturedAtMillis,
                reliableLocationAccuracyM = checkpoint?.lastReliableLocation?.horizontalAccuracyM,
                sessionMode = sessionState.mode.name,
                sessionStartedAtMillis = record?.startedAtMillis,
                sessionHardDeadlineAtMillis = record?.hardDeadlineAtMillis,
                sessionRegistrationExpiresAtMillis = record?.registrationExpiresAtMillis,
                sessionDeliveredUpdateCount = record?.deliveredUpdateCount,
                sessionProfileVersion = record?.profileVersion,
                lastSessionStopReason = sessionState.lastStopReason?.name,
                lastSessionFailure = sessionState.lastFailure,
                drivingConfirmed = sessionState.drivingConfirmed,
                drivingReasonCodes = sessionState.drivingReasonCodes,
                vehicleFirstSeenAtMillis = evidence?.vehicleFirstSeenAtMillis,
                lastVehicleEvidenceAtMillis = evidence?.lastVehicleEvidenceAtMillis,
                reliableSampleCount = evidence?.reliableSampleCount,
                maxSpeedMps = evidence?.maxSpeedMps,
                movingSampleCount = evidence?.movement?.movingSampleCount,
                speedAvailableCount = evidence?.movement?.speedAvailableCount,
                speedMissingCount = evidence?.movement?.speedMissingCount,
                derivedMovingSampleCount = evidence?.movement?.derivedMovingSampleCount,
                movementEvidenceRejectReason = evidence?.movement?.rejectReason?.wire,
                movementOutlierCount = evidence?.movement?.outlierCount,
                counters = sessionState.counters,
                recentFixes = qualityHistory.map { entry ->
                    FixQuality(
                        atMillis = entry.sample.atMillis,
                        accuracyM = entry.sample.horizontalAccuracyM,
                        speedMps = entry.sample.speedMps,
                        dropReason = entry.dropReason?.name,
                    )
                },
                transitionEventCount = transitions.size,
                recentTransitions = transitions.map { event ->
                    TransitionSummary(
                        kind = event.kind.wire,
                        atMillis = event.atMillis,
                        deliveryDelayMillis = event.receivedAtMillis - event.atMillis,
                    )
                },
                trace = trace,
            )
        }

        private fun RegistrationStatus.wire(): String = when (this) {
            RegistrationStatus.Unknown -> "unknown"
            RegistrationStatus.Disabled -> "disabled"
            RegistrationStatus.MissingPermission -> "missing_permission"
            is RegistrationStatus.Active -> "active(spec $specVersion)"
            is RegistrationStatus.Failed -> "failed($reason)"
        }
    }
}
